File: Backend/src/main/java/com/example/employeemanagerapp/service/EmployeeService.java
package com.example.employeemanagerapp.service;

import com.example.employeemanagerapp.exceptions.UserNotFoundException;
import com.example.employeemanagerapp.model.Employee;
import com.example.employeemanagerapp.repository.EmployeeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

/**
 * Service class responsible for handling business logic related to employee entities.
 * Interacts with the {@link EmployeeRepository} for database operations.
 */
@Service
public class EmployeeService {

    /**
     * Autowired instance of {@link EmployeeRepository} for database interaction.
     */
    @Autowired
    private EmployeeRepository employeeRepository;

    /**
     * Adds a new employee to the system, generating a random employee code.
     *
     * @param employee The {@link Employee} object representing the employee to be added.
     * @return The newly added employee with a generated employee code.
     */
    public Employee addEmployee(Employee employee) {
        employee.setEmployeeCode(UUID.randomUUID().toString()); // generate random company id
        return employeeRepository.save(employee);
    }

    /**
     * Retrieves a list of all employees in the system.
     *
     * @return A list containing all employees in the system.
     */
    public List<Employee> findAllEmployees() {
        return employeeRepository.findAll();
    }

    /**
     * Updates an existing employee's information in the system.
     *
     * @param employee The {@link Employee} object representing the updated employee information.
     * @return The updated employee.
     */
    public Employee updateEmployee(Employee employee) {
        return employeeRepository.save(employee);
    }

    /**
     * Retrieves a specific employee based on the provided employee ID.
     *
     * @param id The unique identifier of the employee.
     * @return The requested employee.
     * @throws UserNotFoundException if the employee with the given ID is not found.
     */
    public Employee findEmployeeById(Long id) {
        return employeeRepository.findById(id).orElseThrow(() -> new UserNotFoundException("User with id " + id + " was not found!"));
    }

    /**
     * Deletes an employee from the system based on the provided employee ID.
     *
     * @param id The unique identifier of the employee to be deleted.
     */
    @Transactional
    public void deleteEmployee(Long id) {
        employeeRepository.deleteEmployeeById(id);
    }
}

File: Frontend/employee-manager-app/src/app/employee.service.ts
import { Observable } from 'rxjs'
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Employee } from './employee';
import { environment } from '../enviroments/enviroment';


/**
 * Service class responsible for managing employee-related operations with the server.
 * This service interacts with the backend API to perform CRUD operations on employee data.
 *
 * @export
 * @class EmployeeService
 */
@Injectable({
    providedIn: 'root'
})
export class EmployeeService {

  /**
   * The base URL of the API server.
   *
   * @private
   * @type {string}
   * @memberof EmployeeService
   */
    private apiServerUrl = environment.apiBaseUrl;
    
    /**
     * Creates an instance of EmployeeService.
     *
     * @param {HttpClient} http The Angular HTTP client for making HTTP requests.
     * @memberof EmployeeService
     */
    constructor(private http: HttpClient){}

      /**
       * Retrieves a list of all employees from the server.
       *
       * @returns {Observable<Employee[]>} An Observable containing a list of employees.
       * @memberof EmployeeService
       */
    public getEmployees(): Observable<Employee[]> {
        return this.http.get<Employee[]>(`${this.apiServerUrl}/employee/all`);
      }

      /**
       * Adds a new employee to the server.
       *
       * @param {Employee} employee The employee object to be added.
       * @returns {Observable<Employee>} An Observable containing the newly added employee.
       * @memberof EmployeeService
       */
      public addEmployee(employee: Employee): Observable<Employee> {
        return this.http.post<Employee>(`${this.apiServerUrl}/employee/add`, employee);
      }


      /**
       * Updates an existing employee on the server.
       *
       * @param {Employee} employee The updated employee object.
       * @returns {Observable<Employee>} An Observable containing the updated employee.
       * @memberof EmployeeService
       */
      public updateEmployee(employee: Employee): Observable<Employee> {
        return this.http.put<Employee>(`${this.apiServerUrl}/employee/update`, employee);
      }

      /**
       * Deletes an employee from the server based on the provided employee ID.
       *
       * @param {number} employeeId The unique identifier of the employee to be deleted.
       * @returns {Observable<void>} An Observable indicating the success of the operation.
       * @memberof EmployeeService
       */
      public deleteEmployee(employeeId: number): Observable<void> {
         return this.http.delete<void>(`${this.apiServerUrl}/employee/delete/${employeeId}`);
      }
}